### DISCUSSION TITLE
How to introspect an executable object?

### DISCUSSION THREAD

**Discussion by ilg-ul** [mesonbuild/meson#12410] [Q&A] [answered]
For development purposes, I'd like to display the arrays of properties configured in an executable object, with all dependencies collected.

I tried the naive way:

```
  foreach x : unit_test.include_directories
    message(x)
  endforeach
```

but it failed with:

```
platform-native/meson.build:167:43: ERROR: Expecting lparen got eol.
  foreach x : unit_test.include_directories
```

From this message I understand that `unit_test.include_directories` is accepted only on the left side of an assignment, in other words the object properties cannot be read.

Any solution to this? The same question for dependencies objects.

**Comment by QuLogic** [comment, marked as answer]
There are [no such properties on an `executable`](https://mesonbuild.com/Reference-manual_returned_exe.html), neither for reading nor for writing.

If you want to look at the final result, you may want to try [`meson introspect`](https://mesonbuild.com/Commands.html#introspect).

**Comment by mitchgrout** [comment]
As noted in another answer, there is no way _within meson_ to access the include directories. I've had this issue before with respect to setting up custom linters and other build targets. My resolution so far has been to create a wrapper script which consumes the `compile_commands.json` from the build directory, and extract the relevant information from there.

Could you provide some more context on what you need the include directories for?

**Reply by ilg-ul** [reply]
> Could you provide some more context on what you need the include directories for?

The included directories were just an example, in fact I wanted to nicely print all properties, to have a convenient way to visually validate that all includes and compile/link options are properly collected from all dependencies. 

I already print all these properties after creating each dependency, but I do this (it is funny), because objects are immutable, and this forces me to define variables for each property, compose the content of the properties with various if statements to acomodate for various platforms/compilers/etc, create the dependency and, given that the variables are already there, enumerate and print them.

**Comment by ilg-ul** [comment]
I retain that within meson there is no way to access these object properties.

Thank you for your comments.
